File: app/src/main/java/cc/ioctl/opdisplaytuner/ipc/daemon/internal/DispTunerDaemonProxy.java
package cc.ioctl.opdisplaytuner.ipc.daemon.internal;

import androidx.annotation.NonNull;

import java.io.IOException;
import java.util.Collections;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import cc.ioctl.opdisplaytuner.ipc.daemon.IDispTunerDaemon;

public class DispTunerDaemonProxy implements IDispTunerDaemon {

    public interface NativeEventPacket {
    }

    public static class RawIoEventPacket implements NativeEventPacket {
        public byte[] event;
        public byte[] payload;

        public RawIoEventPacket() {
        }

        public RawIoEventPacket(byte[] event, byte[] payload) {
            this.event = event;
            this.payload = payload;
        }
    }

    static class RawHistoryIoEventList {
        public int totalStartIndex;
        public int totalCount;
        public RawIoEventPacket[] events;

        public RawHistoryIoEventList() {
        }

        public RawHistoryIoEventList(int totalStartIndex, int totalCount, RawIoEventPacket[] events) {
            this.totalStartIndex = totalStartIndex;
            this.totalCount = totalCount;
            this.events = events;
        }
    }

    public void dispatchRemoteEvent(NativeEventPacket event) {
        if (event instanceof RawIoEventPacket) {
            IoEventPacket ioEvent = new IoEventPacket((RawIoEventPacket) event);
            for (OnRemoteEventListener listener : mListeners) {
                listener.onIoEvent(ioEvent);
            }
        } else if (event instanceof RemoteDeathPacket) {
            RemoteDeathPacket death = (RemoteDeathPacket) event;
            for (OnRemoteEventListener listener : mListeners) {
                listener.onRemoteDeath(death);
            }
        }
    }

    private final Set<OnRemoteEventListener> mListeners = Collections.synchronizedSet(new HashSet<>());

    @Override
    public void registerRemoteEventListener(@NonNull OnRemoteEventListener listener) {
        Objects.requireNonNull(listener);
        mListeners.add(listener);
    }

    @Override
    public boolean unregisterRemoteEventListener(@NonNull OnRemoteEventListener listener) {
        Objects.requireNonNull(listener);
        return mListeners.remove(listener);
    }

    @Override
    public native boolean isConnected();

    @Override
    public native String getVersionName();

    @Override
    public native int getVersionCode();

    @Override
    public native String getBuildUuid();

    @Override
    public native int getSelfPid();

    @Override
    public native void exitProcess();

    @Override
    public native boolean isDeviceSupported();

    @Override
    public native boolean initHwServiceConnection(@NonNull String[] soPath) throws IOException;

    @Override
    public native boolean isDisplayFeatureHidlServivceAttached();

    @Override
    public native boolean isDimLayerHbmEnabled();

    @Override
    public native void setDimLayerHbmEnabled(boolean enabled);

    @Override
    public native boolean isDimLayerHbmForceEnabled();

    @Override
    public native void setDimLayerHbmForceEnabled(boolean enabled);

    private native RawHistoryIoEventList ntGetHistoryIoEvents(int startIndex, int count);

    @Override
    public native int deviceDriverIoctl0(int request, long arg);

    @NonNull
    @Override
    public native DaemonStatus getDaemonStatus();


    @Override
    public HistoryIoEventList getHistoryIoEvents(int startIndex, int count) {
        RawHistoryIoEventList raw = ntGetHistoryIoEvents(startIndex, count);
        HistoryIoEventList list = new HistoryIoEventList();
        list.totalStartIndex = raw.totalStartIndex;
        list.totalCount = raw.totalCount;
        list.events = new IoEventPacket[raw.events.length];
        for (int i = 0; i < raw.events.length; i++) {
            list.events[i] = new IoEventPacket(raw.events[i]);
        }
        return list;
    }

    @Override
    public native boolean clearHistoryIoEvents();

    @Override
    public native LogEntryRecord[] getLogsPartial(int startIndex, int count);

    public native NativeEventPacket waitForEvent() throws IOException;
}

File: app/src/main/cpp/libappclient/DispTunerDaemonProxy.cpp
//
// Created by kinit on 2021-10-15.
//

#include "DispTunerDaemonProxy.h"

using namespace ipcprotocol;

using Ids = IDispTunerDaemon::TransactionIds;

uint32_t DispTunerDaemonProxy::getProxyId() const {
    return PROXY_ID;
}

TypedLpcResult<std::string> DispTunerDaemonProxy::getVersionName() {
    return invokeRemoteProcedure<std::string>(Ids::getVersionName);
}

TypedLpcResult<int> DispTunerDaemonProxy::getVersionCode() {
    return invokeRemoteProcedure<int>(Ids::getVersionCode);
}

TypedLpcResult<std::string> DispTunerDaemonProxy::getBuildUuid() {
    return invokeRemoteProcedure<std::string>(Ids::getBuildUuid);
}

TypedLpcResult<int> DispTunerDaemonProxy::getSelfPid() {
    return invokeRemoteProcedure<int>(Ids::getSelfPid);
}

TypedLpcResult<void> DispTunerDaemonProxy::exitProcess() {
    return invokeRemoteProcedure<void>(Ids::exitProcess);
}

TypedLpcResult<bool> DispTunerDaemonProxy::isDeviceSupported() {
    return invokeRemoteProcedure<bool>(Ids::isDeviceSupported);
}

TypedLpcResult<bool> DispTunerDaemonProxy::isHwServiceConnected() {
    return invokeRemoteProcedure<bool>(Ids::isHwServiceConnected);
}

TypedLpcResult<bool> DispTunerDaemonProxy::attachHwHidlService(const std::vector<std::string> &soPath) {
    return invokeRemoteProcedure<bool, const std::vector<std::string> &>(Ids::attachHwHidlService, soPath);
}

TypedLpcResult<IDispTunerDaemon::HistoryIoOperationEventList>
DispTunerDaemonProxy::getHistoryIoOperations(uint32_t start, uint32_t length) {
    return invokeRemoteProcedure<IDispTunerDaemon::HistoryIoOperationEventList, const uint32_t &, const uint32_t &>(
            Ids::getHistoryIoOperations, start, length);
}

TypedLpcResult<bool> DispTunerDaemonProxy::clearHistoryIoEvents() {
    return invokeRemoteProcedure<bool>(Ids::clearHistoryIoEvents);
}

TypedLpcResult<IDispTunerDaemon::DaemonStatus> DispTunerDaemonProxy::getDaemonStatus() {
    return invokeRemoteProcedure<IDispTunerDaemon::DaemonStatus>(Ids::getDaemonStatus);
}

TypedLpcResult<int> DispTunerDaemonProxy::deviceDriverIoctl0(uint32_t request, uint64_t arg) {
    return invokeRemoteProcedure<int, const uint32_t &, const uint64_t &>(Ids::deviceDriverIoctl0, request, arg);
}


TypedLpcResult<std::vector<LogEntryRecord>> DispTunerDaemonProxy::getLogsPartial(uint32_t startIndex, uint32_t count) {
    return invokeRemoteProcedure<std::vector<LogEntryRecord>, const uint32_t &, const uint32_t &>(
            Ids::getLogsPartial, startIndex, count);
}

TypedLpcResult<bool> DispTunerDaemonProxy::isOpDimLayerHbmEnabled() {
    return invokeRemoteProcedure<bool>(Ids::isOpDimLayerHbmEnabled);
}

TypedLpcResult<void> DispTunerDaemonProxy::setOpDimLayerHbmEnabled(bool enabled) {
    return invokeRemoteProcedure<void, const bool &>(Ids::setOpDimLayerHbmEnabled, enabled);
}

TypedLpcResult<bool> DispTunerDaemonProxy::isOpDimLayerHbmForceEnabled() {
    return invokeRemoteProcedure<bool>(Ids::isOpDimLayerHbmForceEnabled);
}

TypedLpcResult<void> DispTunerDaemonProxy::setOpDimLayerHbmForceEnabled(bool enabled) {
    return invokeRemoteProcedure<void, const bool &>(Ids::setOpDimLayerHbmForceEnabled, enabled);
}
